### PR TITLE
Model contribution guide

### PR DESCRIPTION
**Pull request by laserkelvin**
This PR adds notes on how to test functionality for new model contributions, and provides a working example for the PyG flavor of EGNN.

Some valuable feedback from review would be whether the instructions are clear, and whether using `pytest` makes the bar to entry a little too high.

### CODE REVIEW

File: matsciml/models/README.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        batch_size=8,
+        dset_kwargs={"transforms": transforms},
+    )
+    # dummy initialization
+    dm.setup("fit")
+    loader = dm.train_dataloader()
+    batch = next(iter(loader))
+    # run the model without gradient tracking
+    with torch.no_grad():
+        embeddings = egnn_architecture(batch)
```
Comment (melo-gonzo) [inline matsciml/models/README.md:179]: `egnn_architecture` should be `model_fixture` I believe.

File: matsciml/models/README.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
+# Model implementations
+
+At a high level, model implementations in Open MatSciML are expected to emit a system-level embedding; i.e. $B$
```
Comment (melo-gonzo) [inline matsciml/models/README.md:3]: All the B's and D's could be consolidated here for clarity: "... emit DxB embeddings where dimension D is some hyper parameter of your model and B is the minibatch size."

File: matsciml/models/README.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        new_node_feats = self.aggregate(graph, messages)
+        return new_node_feats
+
+
+class AmazingModel(AbstractPyGModel):
+    def _forward(...):
+        for block in self.blocks:
+            node_feats = block(graph, node_feats, edge_feats)
+        graph_feats = readout(node_feats)
+        return graph_feats
```
Comment (melo-gonzo) [inline matsciml/models/README.md:91]: Shouldn't our AmazingModel be returning an `Embeddings` object?

File: matsciml/models/README.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+### PyG models
+
+Ideally, message passing layers implemented in PyG should inherit from the `torch_geometric.nn.MessagePassing` class, and implement
+the corresponding `message`, `propagate`, etc. functions as appropriate.
+
+### Point cloud models
+
+Models that operate on point clouds directly are not necessarily as complex as graph architectures,
+as they do not need to rely on framework abstractions to perform message passing. Tt is still best
```
Comment (melo-gonzo) [inline matsciml/models/README.md:111]: Small typo (Tt -> It).

File: matsciml/models/README.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+For a hypothetical new model, `NewModel` implemented with PyTorch Geometric, the
+test below automatically tests against datasets contained in the registry,
+loads a batch, then runs the data through the model. To adapt it to your own
+architecture, you nominally just need to swap out `model_fixture`.
+
+```python
+import pytest
+
+from matsciml import datasets
+from matsciml.datases.transforms import PeriodicPropertiesTransform, PointCloudToGraphTransform
```
Comment (melo-gonzo) [inline matsciml/models/README.md:137]: Small typo :`.datases` -> `.datasets`

File: matsciml/models/README.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+from matsciml.common.registry import registry
+from matsciml.models import NewModel
```
Comment (melo-gonzo) [inline matsciml/models/README.md:139]: For completeness, we can add:
```python
import torch

from matsciml.lightning.data_utils import MatSciMLDataModule
```

**Review by melo-gonzo — CHANGES_REQUESTED** [review]
Looks great overall! Made a few comments of small things to update. I 'followed along' and implemented `CarmeloNet` (SOTA in no benchmarks currently), and this all seemed to flow nicely.

File: matsciml/models/README.md (hunk 3 above)
Comment (laserkelvin) [reply matsciml/models/README.md:91]: Very good catch, I literally just grabbed it from out of date docs!

File: matsciml/models/README.md (hunk 4 above)
Comment (laserkelvin) [reply matsciml/models/README.md]: Fixed in 9cf499

File: matsciml/models/README.md (hunk 5 above)
Comment (laserkelvin) [reply matsciml/models/README.md]: Fixed in df3a1e9

File: matsciml/models/README.md (hunk 6 above)
Comment (laserkelvin) [reply matsciml/models/README.md:140]: Fixed in 45c404c

File: matsciml/models/README.md (hunk 1 above)
Comment (laserkelvin) [reply matsciml/models/README.md]: Fixed in 6a5c2ca

File: matsciml/models/README.md (hunk 2 above)
Comment (laserkelvin) [reply matsciml/models/README.md]: Fixed in c63e762

File: matsciml/models/README.md (hunk 3 above)
Comment (laserkelvin) [reply matsciml/models/README.md]: Fixed in c63e762

**Comment by melo-gonzo** [comment]
Thanks for the updates. My only other comments are that I think the pytest bits should remain, and be required. Ultimately, it's not too difficult to implement at least some minimal tests, especially with examples already in the repo. Let's leave this up for a day or so to see if any more feedback comes in.

**Comment by laserkelvin** [comment]
We've left this up for a bit now - what do you think @melo-gonzo ?

**Comment by melo-gonzo** [comment]
I'm good to have this merged in. Was waiting to see if there was any other feedback. I thinnk the silence means everyone is happy. Approving!

**Review by melo-gonzo — APPROVED** [review]

**Pull request merged**